thirdparty/cmdconfig/commands/cmdeval/cmdeval.go

```diff
@@ -255,16 +258,47 @@ func (r *EvalFnRunner) SaveFnToKptfile() {
 	case "validator":
 		kf.Pipeline.Validators, usrMsg = r.updateFnList(kf.Pipeline.Validators)
 	}
+
+	mutatedKfAsYNode, err := r.preserveCommentsAndFieldOrder(kf)
+	if err != nil {
+		pr.Printf("function is not added to Kptfile: %v\n", err)
+	}
+
 	// When saving function to Kptfile, the functionConfig should be the relative path
 	// to the kpt package, not the relative path to the current working dir.
 	// error handling are ignored since they have been validated in preRunE.
-	if err = kptfileutil.WriteFile(r.RunFns.Path, kf); err != nil {
+	if err := kptfileutil.WriteFile(r.RunFns.Path, mutatedKfAsYNode); err != nil {
 		pr.Printf("function is not added to Kptfile: %v\n", err)
 		return
 	}
 	pr.Printf(usrMsg)
 }
 
+// preserveCommentsAndFieldOrder syncs the mutated Kptfile with the original to preserve
+// comments and field order, and returns the result as a yaml Node
+func (r *EvalFnRunner) preserveCommentsAndFieldOrder(kf *kptfile.KptFile) (*yaml.Node, error) {
+	kfAsRNode, err := yaml.ReadFile(filepath.Join(r.RunFns.Path, kptfile.KptFileName))
+	if err != nil {
+		return nil, fmt.Errorf("could not read Kptfile: %v", err)
+	}
+	mutatedKfAsBytes, err := yaml.Marshal(kf)
+	if err != nil {
+		return nil, fmt.Errorf("could not Marshal Kptfile into bytes: %v", err)
+	}
+	mutatedKfAsRNode, err := yaml.Parse(string(mutatedKfAsBytes))
+	if err != nil {
+		return nil, fmt.Errorf("could not parse Kptfile: %v", err)
+	}
+	// preserve comments and sync field order
+	if err := comments.CopyComments(kfAsRNode, mutatedKfAsRNode); err != nil {
+		return nil, fmt.Errorf("could not preserve Kptfile comments: %v", err)
+	}
+	if err := order.SyncOrder(kfAsRNode, mutatedKfAsRNode); err != nil {
+		return nil, fmt.Errorf("could not preserve Kptfile field order %v", err)
+	}
```

Comment:
Cool for SyncOrder.
In the cases that 1. we append a new function to mutatedKfAsRNode 2. replace a function with different args in  mutatedKfAsRNode, what order change shall we expect?

Reply:
In general what I've seen in kustomize is that the order of map nodes sometimes changes after going through encoding/decoding. I'm not sure if there will actually be any order changes in either case that you described, but I saw this SyncOrder in kyaml is used when RNodes are mutated, so thought it was best to include it here too.